### DISCUSSION TITLE
Ability to somehow retrieve the current tab from ui.tab_panels

### DISCUSSION THREAD

**Discussion by mbrulatout** [zauberzeug/nicegui#2314] [Ideas / Feature Requests]
Hello,

First of all, your project is awesome :)

I have the following DX need : create tabs in a popup with a different form in each tab, on confirmation, retrieve the fields values from the current tab

For this, I have this piece of code (work-related, had to be stripped out of actual code, sorry) and I've encountered two "issues"

 ```python
    with ui.dialog() as dialog, ui.card():
        with ui.tabs() as tabs:
            tab1 = ui.tab("TAB1")
            tab2 = ui.tab("TAB2")

        with ui.tab_panels(tabs, value="TAB1") as tabs_panel: # ISSUE 1
            with ui.tab_panel(tab1):
                # field1 = ui.input(...)
            with ui.tab_panel(tab2):
                # field2 = ui.input(...)
        with ui.row():
            ui.button("Save", on_click=lambda: dialog.submit("Yes"))
            ui.button("Cancel", on_click=lambda: dialog.submit(None))

        if await dialog:
            if tabs_panel.value == tab1._props["name"]:  # ISSUE 2
                # do something with field1
            else:
                # do something with field2            
```

Issue 1: While I can set the tab_panels value to tab1 (ui.tab object), the behavior becomes inconsistent as `tab_panels.value` is tab1 at first then after switching tabs back and forth, `tab_panels.value` is `"TAB1"`. This forces me to set the value to the tab title at first to get a not-so-complex if/else at the end.

Issue 2: Somehow related to issue 1, I didn't find any convenient way to actually retrieve the selected tab. I need to make sure, through issue 1 mitigation, that the tabs_panel.value is always the selected tab title, but then the `ui.tab` class doesn't expose directly its name and I need to access the internal property `_props["name"]` which seems like a not-so-great DX.

I'm feeling that being able to somehow do `if tabs_panel.selected == tab1` and not set `value="TAB1"` would be a nice QoL improvement

Appreciate the help !

**Comment by stepheweffie** [comment]
Do one thing at a time from here - 

```py
from nicegui import ui


with ui.dialog() as dialog, ui.card():
    with ui.tabs() as tabs:
        tab1 = ui.tab("TAB1")
        tab2 = ui.tab("TAB2")

    with ui.tab_panels(tabs, value=tab1) as tabs_panel:  # ISSUE 1
        with ui.tab_panel(tab1):
            with ui.column().classes('items-center'):
                field1 = ui.input('field1')
        with ui.tab_panel(tab2):
            with ui.column().classes('items-center'):
                field2 = ui.input('field2')

    with ui.row():
        ui.button("Save", on_click=lambda: dialog.submit("Yes"))
        ui.button("Cancel", on_click=lambda: dialog.submit(None))

ui.button('Open a dialog', on_click=dialog.open)

ui.run()
```

What do you want do with this? You want to get the value of the input field in the selected tab when the "save" button is clicked?

**Reply by mbrulatout** [reply]
Exactly, when I click "Save", if i'm on TAB1, I want to get `field1`, else ("TAB2"), get `field2`

Something like if `tabs.panel.selected == tab1`

**Comment by falkoschindler** [comment]
Hi @mbrulatout, thanks for bringing this up!

1. You noticed that `ui.tab_panels(tabs, value=tab1)` initializes the value with a `ui.tab` object, which is replaced with strings after the user switches tabs. This is indeed inconvenient.

    We could translate `tab1` into its name "TAB1". But this could also feel inconsistent. Or should we restrict the `value` parameter to strings only and introduce a new parameter `tab` for setting the current `ui.tab` object? 🤔

2. To compare a `value` with the name of a tab, accessing the private `_props` dictionary is very inconvenient. We should add a public readonly `name` property. 👍

**Reply by mbrulatout** [reply]
Thanks for your answer.
> introduce a new parameter tab for setting the current ui.tab object
1. As long as you could also provide the ability to **get** the `tab` too (`current_tab`, `selected_tab`, or something like that would be clearer IMO), this :arrow_up:  would be a great solution



2. :100: 

With both 1 and 2, people would be able to get/set the current tab, and to compare either to a tab object or its name. :fire:

**Reply by falkoschindler** [reply]
Great idea! So we basically need:

- [ ] a `tab` parameter for `ui.tabs` and `ui.tab_panels` (replaces the `TabPanel` option for `value`),
- [ ] a `tab` property for `ui.tabs` and `ui.tab_panels` to get the current `ui.tab` based on the current `value`,
- [ ] a setter for the `tab` properties to set the `value` based on a given tab (not strictly necessary, but might be handy), and
- [ ] a readonly `name` property for `ui.tab` and `ui.tab_panels`.

For backward-compatibility it would also be nice to add an automatic conversion from `ui.tab` to `str` (with a deprecation warning), in case existing user code creates tabs or tab panels with a tab object as initial value.

Would you like to contribute a pull request, @mbrulatout? 🙂
